lib/tasks/sync_hosts_with_teamdynamix.rake

```diff
@@ -11,26 +11,77 @@ desc <<-DESC.strip_heredoc.squish
     * rake hosts:sync_with_teamdynamix[hostname1,hostname2,..,hostnameX]
 DESC
 namespace :hosts do
-  task :sync_with_teamdynamix, [:hostnames] => :environment do |_task, args|
-    hostnames = Array.wrap(args.hostnames) + args.extras
-    if hostnames.present?
-      hosts = hostnames.collect { |hostname| Host.find_by(name: hostname) }
-    else
-      hosts = Host.all
-    end
-    error_count = 0
-
-    hosts.each do |host|
-      td_api = host.td_api
-      asset_id = host.teamdynamix_asset_id
-      action = asset_id && td_api.asset_exist?(asset_id) ? :update : :create
-      begin
-        td_api.send("#{action}_asset", host)
-      rescue StandardError => e
-        puts "\n\tS.No | hostname | action | result" if error_count.zero?
-        error_count += 1
-        puts "\t#{error_count} | #{host.name} | #{action} | Failed(#{e.message})"
+  task :sync_with_teamdynamix => :environment do |_task|
+    @td_api = TeamdynamixApi.instance
+    @errors = []
+    @hosts_synced = []
+
+    sync_existing_assets_to_hosts
+
+    create_assets_for_unmapped_hosts
+
+    print_summary
+  end
+
+  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+  def sync_existing_assets_to_hosts
+    # sending empty search param to get all the assets in TD at once
+    # Note: due to a limitation(possibly a bug) in TD search asset api
+    #       each and every search payload returns all the assets.
+    @teamdynamix_assets = @td_api.search_asset({})
+
+    @teamdynamix_assets.each do |asset|
+      asset_id = asset['ID']
+      # WHEN Asset is already mapped to a host
+      # THEN update the asset in TD as per current configuration
+      host = Host.find_by(teamdynamix_asset_id: asset_id)
+      if host.present?
+        @hosts_synced << host.id
+        @td_api.update_asset(host)
+        next
       end
+      hosts = Host.where('name IN (?)', [asset['Name'], asset['SerialNumber']])
```

Comment:
There should be a ruby way to do this in rails without needing to resort to sql strings

Reply:
I will check.

Reply:
You might check if simply passing an array as a key with a hash works.
Also arel comes to mind.

Reply:
https://robots.thoughtbot.com/using-arel-to-compose-sql-queries

That has an example of where with array